GVFS/GVFS.Common/ProductUpgrader.cs

```diff
@@ -0,0 +1,489 @@
+                            $"Make sure you have write permissions to directory {rootDirectoryPath} and run {GVFSConstants.UpgradeVerbMessages.GVFSUpgradeConfirm} again.");
+                        this.TraceException(e, nameof(this.TrySetupToolsDirectory), $"Error copying {toolPath} to {destinationPath}.");
+                        break;
+                    }
+                    catch (IOException e)
+                    {
+                        error = "File copy error - " + e.Message;
+                        this.TraceException(e, nameof(this.TrySetupToolsDirectory), $"Error copying {toolPath} to {destinationPath}.");
+                        break;
+                    }
+                }
+
+                upgraderToolPath = string.IsNullOrEmpty(error) ? Path.Combine(toolsDirectoryPath, UpgraderToolName) : null;
+                return string.IsNullOrEmpty(error);
+            }
+
+            upgraderToolPath = null;
+            error = exception.Message;
+            this.TraceException(exception, nameof(this.TrySetupToolsDirectory), $"Error creating upgrade tools directory {toolsDirectoryPath}.");
+            return false;
+        }
+
+        public bool TryCleanup(out string error)
+        {
+            error = string.Empty;
+            if (this.newestRelease == null)
+            {
+                return true;
+            }
+
+            foreach (Asset asset in this.newestRelease.Assets)
+            {
+                Exception exception;
+                if (!this.TryDeleteDownloadedAsset(asset, out exception))
+                {
+                    error += $"Could not delete {asset.LocalPath}. {exception.ToString()}." + Environment.NewLine;
+                }
+            }
+
+            if (!string.IsNullOrEmpty(error))
+            {
+                error.TrimEnd(Environment.NewLine.ToCharArray());
+                return false;
+            }
+
+            error = null;
+            return true;
+        }
+        
+        public virtual bool TryLoadRingConfig(out string error)
+        {
+            string gitPath = GVFSPlatform.Instance.GitInstallation.GetInstalledGitBinPath();
+            LocalGVFSConfig localConfig = new LocalGVFSConfig();
+
+            string ringConfig = null;
+            if (localConfig.TryGetConfig(GVFSConstants.LocalGVFSConfig.UpgradeRing, out ringConfig, out error, this.tracer))
+            {
+                RingType ringType;
+
+                if (Enum.TryParse(ringConfig, ignoreCase: true, result: out ringType) &&
+                    Enum.IsDefined(typeof(RingType), ringType) &&
+                    ringType != RingType.Invalid)
+                {
+                    this.Ring = ringType;
+                    error = null;
+                    return true;
+                }
+                else
+                {
+                    error = "Invalid upgrade ring `" + ringConfig + "` specified in gvfs config.";

```

Comment:
This is expected behavior today - When no ring is specified, or the ring that is specified is mis-spelled, it is treated as an error. We show the error and advice user on how to set the right ring config. This behavior is different from "None" ring, which is an acceptable ring setting.

Reply:
>> Otherwise everyone's first interaction with this command will be an error.

Makes sense. Removed ERROR. Have attached screenshot with the new behavior in a comment below.